README
NAME
    Path::Class::Rule - Iterative, recursive file finder with Path::Class

VERSION
    version 0.017

SYNOPSIS
      use Path::Class::Rule;

      my $rule = Path::Class::Rule->new; # match anything
      $rule->file->size(">10k");         # add/chain rules

      # iterator interface
      my $next = $rule->iter( @dirs );
      while ( my $file = $next->() ) {
        ...
      }

      # list interface
      for my $file ( $rule->all( @dirs ) ) {
        ...
      }

DESCRIPTION
    This module iterates over files and directories to identify ones
    matching a user-defined set of rules. The API is based heavily on
    File::Find::Rule, but with more explicit distinction between matching
    rules and options that influence how directories are searched. A
    "Path::Class::Rule" object is a collection of rules (match criteria)
    with methods to add additional criteria. Options that control directory
    traversal are given as arguments to the method that generates an
    iterator.

    As of version 0.016, this is now a thin subclass of Path::Iterator::Rule
    that operates on and returns Path::Class objects instead of bare file
    paths.

    See that module for details on features and usage.

EXTENDING
    This module may be extended in the same way as "Path::Iterator::Rule",
    but test subroutines receive "Path::Class" objects instead of strings.

    Consider whether you should extend "Path::Iterator::Rule" or
    "Path::Class::Rule". Extending this module specifically is recommended
    if your tests rely on having a "Path::Class" object.

LEXICAL WARNINGS
    If you run with lexical warnings enabled, "Path::Iterator::Rule" will
    issue warnings in certain circumstances (such as a read-only directory
    that must be skipped). To disable these categories, put the following
    statement at the correct scope:

      no warnings 'Path::Iterator::Rule';

SUPPORT
  Bugs / Feature Requests
    Please report any bugs or feature requests through the issue tracker at
    <https://github.com/dagolden/path-class-rule/issues>. You will be
    notified automatically of any progress on your issue.

  Source Code
    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    <https://github.com/dagolden/path-class-rule>

      git clone git://github.com/dagolden/path-class-rule.git

AUTHOR
    David Golden <dagolden@cpan.org>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2011 by David Golden.

    This is free software, licensed under:

      The Apache License, Version 2.0, January 2004



Changes
Revision history for Path-Class-Rule

0.017     2013-01-23 14:44:04 America/New_York

  [CHANGES]

    - Updated for internal changes to Path::Iterator::Rule 0.002

0.016     2013-01-22 06:28:03 America/New_York

  [CHANGES]

    - Reimplmented as a thin subclass of Path::Iterator::Rule

0.015     2013-01-21 14:21:33 America/New_York

  [DOCUMENTATION]

    - Oops.  Left off a doc fix from the master repo.  Merged now.

0.014     2013-01-21 14:19:04 America/New_York

  [OPTIMIZATION]

    - Use Path::Class's basename() internally,, resulting in a 50%
      speedup on a test case of finding all .pm files in @INC; requires
      Path::Class 0.20.

0.013     2012-12-13 11:08:42 America/New_York

  [DOCUMENTATION]

    - The method for all perl file types is 'perl_file', not 'perl_files'
      [rt.cpan.org #81945]

0.012     2012-10-26 21:51:27 America/New_York

  [BUG FIXES]

    - Rule cloning now does deep clone with Data::Clone [rt.cpan.org #80426]

0.011     2012-03-15 09:55:33 EST5EDT

  [BUG FIXES]

    - Fixed major bugs in depth-first traversal relating to proper ordering
      and incorrectly returning directories that didn't match search
      criteria

0.010     2011-12-03 20:17:55 EST5EDT

  [OTHER]

    - Removed 'parent' and 'File::Copy' as dependencies.  Former wasn't
      really needed and latter was causing problems on 5.10.0.

0.009     2011-12-02 04:38:05 EST5EDT

  [BUG FIXES]

    - Fix t/error_handler.t regex to prevent errors from backslashes in
      paths on Windows

0.008     2011-11-30 09:54:03 EST5EDT

  [DOCUMENTATION]

    - Added File::Set to "See Also" section

  [OTHER]

    - Removed namespace::autoclean as a dependency because it pulls in
      the whole Moose stack.  [rt.cpan.org #72841, patch by Lars Dɪᴇᴄᴋᴏᴡ]

0.007     2011-10-26 23:50:57 America/New_York

  [BUG FIXES]

    - Fixed Win32 test failures

    - Fixed regexp handling on Perl 5.10

0.006     2011-10-19 21:40:01 America/New_York

  [BUG FIXES]

    - Fixed incompability with Perl < v5.14 [rt.cpan.org #71400]

  [DOCUMENTATION]

    - Clarified caveats about taint mode.  Path::Class::Rule runs under
      taint mode; it offers no options to automatically untaint data

0.005     2011-09-14 16:58:25 America/New_York

  [NEW FEATURES]

    - Added 'iname' rule method for case insensitive name matching

    - Added 'skip' logic method for easy custom pruning

    - Added 'error_handler' option to iteration method

    - Added lexical warnings category

  [BUG FIXES]

    - Directories that are not readable will warn and continue instead
      of throwing an error

  [DOCUMENTATION]

    - Various clarifications and typo fixes

0.004     2011-09-14 06:48:13 America/New_York

  [DOCUMENTATION]

    - Added additional modules to SEE ALSO commentary

0.003     2011-09-13 21:05:43 America/New_York

  [NEW FEATURES]

    - Added 'skip_vcs' (and individual VCS skipping rule methods)

    - Added 'perl_files' (and similar rule methods)

    - Added 'shebang' rule method

0.002     2011-09-12 14:17:13 America/New_York

  [NEW FEATURES]

    - Added 'min_depth' and 'max_depth' rule methods

    - Added 'dangling' rule method to match dangling symlinks

    - Added 'loop_safe' option to control infinite loop protection
      (default is true)

    - Rule callbacks receive a hash reference for storing optional state

  [BUG FIXES]

    - Bumped Number::Compare prerequisite version to 0.02 to
      avoid uninitialized warnings when comparing numbers without
      a magnitude.

0.001     2011-09-11 22:40:44 America/New_York

    - Initial release



lib/Path/Class/Rule.pm
use strict;
use warnings;

package Path::Class::Rule;
# ABSTRACT: Iterative, recursive file finder with Path::Class
our $VERSION = '0.017'; # VERSION

use Path::Iterator::Rule 0.002; # new _children API
our @ISA = qw/Path::Iterator::Rule/;

use Path::Class;
use namespace::clean;

sub _objectify {
    my $self = shift;
    my $path = "" . shift;
    return -d $path ? dir($path) : file($path);
}

sub _children {
    my $self = shift;
    my $path = shift;
    return map { [ $_->basename, $_ ] } $path->children;
}

1;


# vim: ts=4 sts=4 sw=4 et:

__END__

=pod

=head1 NAME

Path::Class::Rule - Iterative, recursive file finder with Path::Class

=head1 VERSION

version 0.017

=head1 SYNOPSIS

  use Path::Class::Rule;

  my $rule = Path::Class::Rule->new; # match anything
  $rule->file->size(">10k");         # add/chain rules

  # iterator interface
  my $next = $rule->iter( @dirs );
  while ( my $file = $next->() ) {
    ...
  }

  # list interface
  for my $file ( $rule->all( @dirs ) ) {
    ...
  }

=head1 DESCRIPTION

This module iterates over files and directories to identify ones matching a
user-defined set of rules.  The API is based heavily on L<File::Find::Rule>,
but with more explicit distinction between matching rules and options that
influence how directories are searched.  A C<Path::Class::Rule> object is a
collection of rules (match criteria) with methods to add additional criteria.
Options that control directory traversal are given as arguments to the method
that generates an iterator.

As of version 0.016, this is now a thin subclass of L<Path::Iterator::Rule>
that operates on and returns L<Path::Class> objects instead of bare file paths.

See that module for details on features and usage.

=head1 EXTENDING

This module may be extended in the same way as C<Path::Iterator::Rule>, but
test subroutines receive C<Path::Class> objects instead of strings.

Consider whether you should extend C<Path::Iterator::Rule> or C<Path::Class::Rule>.
Extending this module specifically is recommended if your tests rely on having
a C<Path::Class> object.

=head1 LEXICAL WARNINGS

If you run with lexical warnings enabled, C<Path::Iterator::Rule> will issue
warnings in certain circumstances (such as a read-only directory that must be
skipped).  To disable these categories, put the following statement at the
correct scope:

  no warnings 'Path::Iterator::Rule';

=for :stopwords cpan testmatrix url annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata placeholders metacpan

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<https://github.com/dagolden/path-class-rule/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/dagolden/path-class-rule>

  git clone git://github.com/dagolden/path-class-rule.git

=head1 AUTHOR

David Golden <dagolden@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2011 by David Golden.

This is free software, licensed under:

  The Apache License, Version 2.0, January 2004

=cut


Makefile.PL

use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Iterative, recursive file finder with Path::Class",
  "AUTHOR" => "David Golden <dagolden\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Config" => 0,
    "Exporter" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "File::pushd" => 0,
    "Test::Deep" => 0,
    "Test::More" => "0.92",
    "autodie" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Path-Class-Rule",
  "EXE_FILES" => [],
  "LICENSE" => "apache",
  "NAME" => "Path::Class::Rule",
  "PREREQ_PM" => {
    "Path::Class" => 0,
    "Path::Iterator::Rule" => "0.002",
    "namespace::clean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.017",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



